pyresample/bucket/__init__.py

```diff
@@ -211,35 +213,42 @@ def _call_bin_statistic(self, statistic_method, data, fill_value=None, skipna=No
 
         # sort idxs and weights based on weights
         if statistic_method == 'min':
-            order = da.map_blocks(np.argsort, weights)
+            order = np.argsort(weights)
         elif statistic_method == 'max':
-            order = da.map_blocks(np.argsort, weights)[::-1]
+            order = np.argsort(weights)[::-1]
+
+        @dask.delayed
+        def get_block_statistic(block, mask_buffer, idxs_sorted, weights_sorted):
+            extended_bins = np.append(block, np.iinfo(block.dtype).max)
+            # use the preallocated buffers
+            mask_buffer[:len(bins), :-1] = (bins[:, None] <= idxs_sorted[None, :]) &\
```

Comment:
I use the preallocated buffers, but it could lead to a different result in each run. That's strange.
If I create the new mask in each run like this:
```
            mask = (block[:, None] <= idxs_sorted[None, :-1]) &\
                          (extended_bins[1:, None] > idxs_sorted[None, :-1])
            mask = np.hstack((mask, np.ones((mask.shape[0], 1), dtype=mask.dtype)))

```
it works well but will raise memory value.

Reply:
Why would it lead to a different result in each run? Can you move the creation of `mask_buffer` into this function? Oh wait, I see the `for block in blocks` now. This isn't really what I had in mind. If you are already doing delayed functions then you should probably just put this whole computation into a delayed function. As I mentioned before this completely removes the chunked/parallel nature of dask, but if it works then that's at least a step in the right direction.

I had said that you need to separate the algorithm into the different components: the parts that operate on per-dask-chunk and the ones that need to operate on the whole array. This `mask_buffer` concept needs to operate on the whole array so the operations using it need to all be in the same function.